In [7]:
from nltk.corpus import wordnet

<b>1) Найти все значения (синсеты) для лексемы plant<b>

In [8]:
bank = wordnet.synsets('bank')
for ss in bank:
    print(ss, ss.definition())

Synset('bank.n.01') sloping land (especially the slope beside a body of water)
Synset('depository_financial_institution.n.01') a financial institution that accepts deposits and channels the money into lending activities
Synset('bank.n.03') a long ridge or pile
Synset('bank.n.04') an arrangement of similar objects in a row or in tiers
Synset('bank.n.05') a supply or stock held in reserve for future use (especially in emergencies)
Synset('bank.n.06') the funds held by a gambling house or the dealer in some gambling games
Synset('bank.n.07') a slope in the turn of a road or track; the outside is higher than the inside in order to reduce the effects of centrifugal force
Synset('savings_bank.n.02') a container (usually with a slot in the top) for keeping money at home
Synset('bank.n.09') a building in which the business of banking transacted
Synset('bank.n.10') a flight maneuver; aircraft tips laterally about its longitudinal axis (especially in turning)
Synset('bank.v.01') tip laterally
Sy

<b>2) Найти определение для лексемы bank в значении  (a) "берег" и в значении  (b) "финансовое учреждение"</b>

In [9]:
print(bank[0], bank[0].definition())
print(bank[1], bank[1].definition())

Synset('bank.n.01') sloping land (especially the slope beside a body of water)
Synset('depository_financial_institution.n.01') a financial institution that accepts deposits and channels the money into lending activities


<b>3) Найдите два произвольных контекста для слова bank в значениях (a) "берег" и (b) "финансовое учреждение"; продемонстрируйте на них действие алгоритма Леска для разрешения неоднозначности</b>

In [14]:
sent1 = 'No visit to Paris would be complete without a stroll along the banks of the River Seine.'
sent2 = 'Welcome to Bank of America, home for all of your financial needs.'
sent1_tokens = [word.strip('.,') for word in sent1.split(' ')]
sent2_tokens = [word.strip('.,') for word in sent2.split(' ')]
print(sent1_tokens)
print(sent2_tokens)

['No', 'visit', 'to', 'Paris', 'would', 'be', 'complete', 'without', 'a', 'stroll', 'along', 'the', 'banks', 'of', 'the', 'River', 'Seine']
['Welcome', 'to', 'Bank', 'of', 'America', 'home', 'for', 'all', 'of', 'your', 'financial', 'needs']


In [15]:
from nltk.wsd import lesk
print(lesk(sent1_tokens, 'bank').definition())
print(lesk(sent2_tokens, 'bank').definition())

a slope in the turn of a road or track; the outside is higher than the inside in order to reduce the effects of centrifugal force
a container (usually with a slot in the top) for keeping money at home


В первом случае значение определено почти правильно, хотя оно довольно странно сформулировано. Во втором случае значение абсолютно неверно. 

<b>4) Найдите гиперонимы для значения (a) и гиперонимы для значения (b)</b>

In [16]:
for ss in bank[0].hypernyms():
    print(ss, ss.definition())
for ss in bank[1].hypernyms():
    print(ss, ss.definition())

Synset('slope.n.01') an elevated geological formation
Synset('financial_institution.n.01') an institution (public or private) that collects funds (from the public or other institutions) and invests them in financial assets


<b>5) Вычислите наименьшее расстояние между значением bank "берег" и значениями лексемы river, а также bank "банк" и значениями лексемы money.
    
Найти min (d(bank: "берег", river), d(bank: "банк", money)), а также min (d(bank: "берег", money), d(bank: "банк", river))</b>

In [24]:
river = wordnet.synsets('river')
money = wordnet.synsets('money')

def get_dist_sim(ss1, lexeme):
    distances = []
    similarities = []
    for ss2 in lexeme:
        dist = ss1.shortest_path_distance(ss2)
        if dist is not None:
            distances.append(dist)
            sim = ss1.path_similarity(ss2)
            similarities.append(sim)
    return distances, similarities

# min d(bank: "берег", river)
dist1 = get_dist_sim(bank[0], river)[0]
print('min d(bank: "берег", river): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(river[dist1.index(min(dist1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# min d(bank: "берег", money)
dist2 = get_dist_sim(bank[0], money)[0]
print('min d(bank: "берег", money): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(money[dist2.index(min(dist2))].definition()))

# min d(bank: "банк", money)
dist3 = get_dist_sim(bank[1], money)[0]
print('mind (bank: "банк", money): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(money[dist3.index(min(dist3))].definition()))

# min d(bank: "банк", river)
dist4 = get_dist_sim(bank[1], river)[0]
print('min d(bank: "банк", river): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(river[dist4.index(min(dist4))].definition()))

print('min (d(bank: "берег", river), d(bank: "берег", money)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(bank: "банк", money), d(bank: "банк", river)): {}'.format(min(min(dist3), min(dist4))))

min d(bank: "берег", river): 8
closest lemma definition: a large natural stream of water (larger than a creek)

min d(bank: "берег", money): 11
closest lemma definition: the most common medium of exchange; functions as legal tender

mind (bank: "банк", money): 11
closest lemma definition: the most common medium of exchange; functions as legal tender

min d(bank: "банк", river): 12
closest lemma definition: a large natural stream of water (larger than a creek)

min (d(bank: "берег", river), d(bank: "берег", money)): 8
min (d(bank: "банк", money), d(bank: "банк", river)): 11


Кратчайшее расстояние от берега до river - 8, до money - 12. Получается, что до river расстояние короче - логично.
Кратчайшее расстояние от банка до money - 11, до leaf - 12. Получается, до money расстояние короче, но странно, что только на 1.

Если имелось в виду посчитать не расстояние, а сходство путей на основе кратчайшего расстояния (минимальный путь --> максимальное сходство), вот оно:

In [23]:
# max s(bank: "берег", river)
sim1 = get_dist_sim(bank[0], river)[1]
print('max s(bank: "берег", river): {}'.format(max(sim1)))
print('closest lemma definition: {}\n'.format(river[sim1.index(max(sim1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# max s(bank: "берег", money)
sim2 = get_dist_sim(bank[0], money)[1]
print('max s(bank: "берег", money): {}'.format(max(sim2)))
print('closest lemma definition: {}\n'.format(money[sim2.index(max(sim2))].definition()))

# max s(bank: "банк", money)
sim3 = get_dist_sim(bank[1], money)[1]
print('max s(bank: "банк", money): {}'.format(max(sim3)))
print('closest lemma definition: {}\n'.format(money[sim3.index(max(sim3))].definition()))

# max s(bank: "банк", river)
sim4 = get_dist_sim(bank[1], river)[1]
print('max s(bank: "банк", river): {}'.format(max(sim4)))
print('closest lemma definition: {}\n'.format(river[sim4.index(max(sim4))].definition()))

print('max (s(bank: "берег", river), s(bank: "берег", money)): {}'.format(max(max(sim1), max(sim2))))
print('max (s(bank: "банк", money), s(bank: "банк", river)): {}'.format(max(max(sim3), max(sim4))))

max s(bank: "берег", river): 0.1111111111111111
closest lemma definition: a large natural stream of water (larger than a creek)

max s(bank: "берег", money): 0.08333333333333333
closest lemma definition: the most common medium of exchange; functions as legal tender

max s(bank: "банк", money): 0.08333333333333333
closest lemma definition: the most common medium of exchange; functions as legal tender

max s(bank: "банк", river): 0.07692307692307693
closest lemma definition: a large natural stream of water (larger than a creek)

max (s(bank: "берег", river), s(bank: "берег", money)): 0.1111111111111111
max (s(bank: "банк", money), s(bank: "банк", river)): 0.08333333333333333


Результат тот же: берег ближе к river, банк ближе к money.

<b>6) Вычислить двумя разными способами расстояние:
d(bank: "банк", bank account), d(bank: "берег", shore) и d(soup, alien)
Есть ли разница в расстояниях? Какое из расстояний, по Вашему мнению, в лучшей степени отражает интуитивное представление о семантчиеской близости слов?</b>

<i>Здесь я не очень поняла алгоритм выбора слов. Я решила что первая пара - максимально похожие слова, а вторая - максимально далекии по значению. Так как у слова bank я не нашла хороших гипонимов, я решила сравнить 'банк' с банковским счетом, и 'берег' c shore (тоже берег, но морской). А в качестве максимально далекой пары я взяла alien и soup.</i>

In [64]:
account = wordnet.synsets("bank_account")
for ss in account:
    print(ss, ss.definition())

Synset('bank_account.n.01') a fund that a customer has entrusted to a bank and from which the customer can make withdrawals


In [62]:
shore = wordnet.synsets("shore")
for ss in shore:
    print(ss, ss.definition())

Synset('shore.n.01') the land along the edge of a body of water
Synset('shore.n.02') a beam or timber that is propped against a structure to provide support
Synset('shore.v.01') serve as a shore to
Synset('land.v.06') arrive on shore
Synset('prop_up.v.01') support by placing against something solid or rigid


Посчитаем близость с помощью критериев Path Similarity, Leacock-Chodorow Similarity, Wu-Palmer Similarity, Resnik Similarity, Jiang-Conrath Similarity и Lin Similarity. Для последних трех будем использовать Information Content корпуса Brown.

In [67]:
from nltk.corpus import wordnet_ic
ic = wordnet_ic.ic('ic-brown.dat')

print(bank[1].path_similarity(account[0]))

#Leacock-Chodorow Similarity
print(bank[1].lch_similarity(account[0]))

#Wu-Palmer Similarity
print(bank[1].wup_similarity(account[0]))
    
#Resnik Similarity
print(bank[1].res_similarity(account[0], ic))
    
#Jiang-Conrath Similarity
print(bank[1].jcn_similarity(account[0], ic))
    
#Lin Similarity
print(bank[1].lin_similarity(account[0], ic))


0.08333333333333333
1.1526795099383855
0.26666666666666666
0.5962292078977726
1e-300
1.1924584157955451e-300


In [69]:
ic = wordnet_ic.ic('ic-brown.dat')

print(bank[0].path_similarity(shore[0]))

#Leacock-Chodorow Similarity
print(bank[0].lch_similarity(shore[0]))

#Wu-Palmer Similarity
print(bank[0].wup_similarity(shore[0]))
    
#Resnik Similarity
print(bank[0].res_similarity(shore[0], ic))
    
#Jiang-Conrath Similarity
print(bank[0].jcn_similarity(shore[0], ic))
    
#Lin Similarity
print(bank[0].lin_similarity(shore[0], ic))

0.25
2.2512917986064953
0.7272727272727273
5.884681030765606
0.137293462201792
0.6177160279043438


In [44]:
soup = wordnet.synsets('soup', 'n')
for ss in soup:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('soup.n.01') liquid food especially of meat or fish or vegetable stock often containing pieces of solid food
Synset('soup.n.01') []
Synset('soup.n.02') any composition having a consistency suggestive of soup
Synset('soup.n.02') []
Synset('soup.n.03') an unfortunate situation
Synset('soup.n.03') ["we're in the soup now"]


Так как мы хотим взять максимально несвязанные слова, то логично взять значение soup "liquid food especially of meat or fish or vegetable stock often containing pieces of solid food"

In [45]:
alien = wordnet.synsets('alien', 'n')
for ss in alien:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('foreigner.n.01') a person who comes from a foreign country; someone who does not owe allegiance to your country
Synset('foreigner.n.01') []
Synset('stranger.n.01') anyone who does not belong in the environment in which they are found
Synset('stranger.n.01') []
Synset('extraterrestrial_being.n.01') a form of life assumed to exist outside the Earth or its atmosphere
Synset('extraterrestrial_being.n.01') []


Так как мы хотим взять максимально несвязанные слова, то логично взять значение alien "anyone who does not belong in the environment in which they are found" 

In [55]:
'''
на всякий случай выведу все соответствия, хотя можно было бы и не смотреть по всем, потому что
остальные значения довольно странные
'''
for ss1 in soup:
    print(ss1.definition())
    for ss2 in alien:
        print(ss2.definition())
        print(ss1.path_similarity(ss2))

        #Leacock-Chodorow Similarity
        print(ss1.lch_similarity(ss2))

        #Wu-Palmer Similarity
        print(ss1.wup_similarity(ss2))

        #Resnik Similarity
        print(ss1.res_similarity(ss2, ic))

        #Jiang-Conrath Similarity
        print(ss1.jcn_similarity(ss2, ic))

        #Lin Similarity
        print(ss1.lin_similarity(ss2, ic))

liquid food especially of meat or fish or vegetable stock often containing pieces of solid food
a person who comes from a foreign country; someone who does not owe allegiance to your country
0.09090909090909091
1.2396908869280152
0.2857142857142857
0.8017591149538994
0.05727299508402372
0.08411345562516616
anyone who does not belong in the environment in which they are found
0.08333333333333333
1.1526795099383855
0.26666666666666666
0.8017591149538994
0.05891403853702992
0.08631552960174331
a form of life assumed to exist outside the Earth or its atmosphere
0.058823529411764705
0.8043728156701697
0.1111111111111111
-0.0
0.04881961232302059
-0.0
any composition having a consistency suggestive of soup
a person who comes from a foreign country; someone who does not owe allegiance to your country
0.1
1.3350010667323402
0.3076923076923077
0.8017591149538994
0.049729541194259624
0.07385302154474695
anyone who does not belong in the environment in which they are found
0.09090909090909091
1.23

<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
    
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th>Similarity Index</th><th>банк/банковский счет</th><th>bank"берег"/ shore</th><th>liquid food / alien("инопланетянин")</th></tr></thead><tbody>
 <tr><td>Path Similarity</td><td>0.08333333333333333</td><td>0.25</td><td>0.058823529411764705</td></tr>
 <tr><td>Leacock-Chodorow Similarity</td><td>1.1526795099383855</td><td>2.2512917986064953</td><td>0.8043728156701697</td></tr>
 <tr><td>Wu-Palmer Similarity</td><td>0.26666666666666666</td><td>0.7272727272727273</td><td>0.1111111111111111</td></tr>
 <tr><td>Resnik Similarity</td><td>0.5962292078977726</td><td>5.884681030765606</td><td>-0.0</td></tr>
 <tr><td>Jiang-Conrath Similarity</td><td>1e-300</td><td>0.137293462201792</td><td>0.04881961232302059</td></tr>
 <tr><td>Lin Similarity</td><td>1.1924584157955451e-300</td><td>0.6177160279043438</td><td>-0.0</td></tr>
</tbody></table>

Расстояния получились очень разные.
Bank account, хоть и не является гипонимом банка, но в каком-то смысле является частью его услуг. Shore - должен употребляться похожим образом как и river bank, хотя и не польностью ему синонимичен.
Поэтому логично, что берег и shore получили самые высокие результаты по первым четырем параметрам.
При этом bank & shore получили более высокие показатели, чем банк & банковский счет, что тоже логично (контексты последней пары сильнее различаются контекстов первой).

Очень странное число возвращает Jiang-Conrath Similarity на Паре банк/банковский счет. Судя по тому, что я сравнивала, эти результаты сложнее всего как-то интерпретировать. Lin Similarity тоже дает не очень понятные результаты.

Как мне кажется, Leacock-Chodorow Similarity в лучшей степени отражает интуитивное представление о семантчиеской близости слов.



